fendermint/vm/interpreter/src/fvm/query.rs

```diff
@@ -99,3 +96,131 @@ where
         Ok((state, res))
     }
 }
+
+impl<DB> FvmMessageInterpreter<DB>
+where
+    DB: Blockstore + 'static + Send + Sync + Clone,
+{
+    /// Overestimation rate applied to gas to ensure that the
+    /// message goes through in the gas estimation.
+    const GAS_OVERESTIMATION_RATE: f64 = 1.25;
+    /// Default gas premium value. Inferred through a quick search through
+    /// InvokeEVM messages in filfox. The default value is only used if
+    /// the user hasn't specified a gas premium.
+    const DEFAULT_GAS_PREMIUM: u64 = 20000;
+    /// Gas search step increase used to find the optimal gas limit.
+    const GAS_SEARCH_STEP: f64 = 1.2;
+
+    fn estimate_gassed_msg(
+        &self,
+        state: &FvmQueryState<DB>,
+        msg: Box<Message>,
+    ) -> anyhow::Result<Message> {
+        let mut out = (*msg).clone();
+        // estimate the gas limit and assign it to the message
+        // do not reuse the cache
+        let ret = state.call_with_cache(*msg.clone(), false)?;
+        if ret.msg_receipt.exit_code != ExitCode::OK {
+            return Err(anyhow::anyhow!(
+                "message execution failed with error code {}",
+                ret.msg_receipt.exit_code
+            ));
+        }
+        out.gas_limit = (ret.msg_receipt.gas_used as f64 * Self::GAS_OVERESTIMATION_RATE) as u64;
+        if out.gas_premium.is_zero() {
+            // TODO: Instead of assigning a default value here, we should analyze historical
+            // blocks from the current height to estimate an accurate value for this premium.
+            // To achieve this we would need to perform a set of ABCI queries.
+            // In the meantime, this value should be good enough to make sure that the
+            // message is included in a block.
+            out.gas_premium = TokenAmount::from_nano(BigInt::from(Self::DEFAULT_GAS_PREMIUM));
```

Comment:
ethers-rs [does this](https://github.com/gakonst/ethers-rs/blob/ethers-v2.0.8/ethers-providers/src/rpc/provider.rs#L440), do we really have to do it here? This is forcing rewards onto users via estimates. Shouldn't it be up to them and the miners?

Reply:
In Lotus they have a local estimation of fees also in case the user didn't set the `gas_premium`. The reason for that is that you can use Lotus to send non-EVM messages directly and the user may not set the premium. I guess having a constant premium here as a sanity-check should be OK for us because we will rely on `ethers` and other tooling to compute this for us, so we should expect almost all messages to have `gas_premium != 0`. 

Weak opinion weakly held. I can also bypass `gas_premium` here even if it is zero if you think this is the correct way of doing it.

Reply:
I don't know, I didn't think there was a correct way because the FVM happily executes with 0 fees and 0 premiums, and it should be up to the validator to make sure they only include transactions which are profitable, see #29 

Personally I would not set anything that isn't necessary to be set, especially since this value is not returned from the API, so even now I'm not sure what effect we have by setting it.

Reply:
After reading the spec and tinkering with it, I think you are right, is better to just estimate the `gas_limit` for now, and then compute the `gas_premium` and `gas_fee_cap` optimally once we figure out the `base_fee` story. Unless you are against, let's defer for the future.